### PR TITLE
Add Class-Incremental Sampler

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

Class-Incremental Sampler from
https://github.com/openvinotoolkit/training_extensions/blob/develop/src/otx/algorithms/common/adapters/torch/dataloaders/samplers/cls_incr_sampler.py

While i've verified that it works as intended, this sampler usage is a bit of a pain for CLI users.
The new_classes and old_classes have to be manually entered by the user, which is something we'll be thinking about improving.

image

I've simply checked the above results using a dataset called coco_incr, and we need to do more experiments to verify them.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by kprokofi — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]
Still, I don't prefer the manual typing for the old_classes and new_classes, however, let's consider the painpoint at the near future

File: src/otx/algo/samplers/class_incremental_sampler.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                    [
+                        self.new_indices[torch.randint(0, len(self.new_indices), (1,), generator=generator)],
+                        # random_generator.choice(self.new_indices, 1),
+                        self.old_indices[
+                            torch.randint(0, len(self.old_indices), (self.old_new_ratio,), generator=generator)
+                        ],
+                        # random_generator.choice(self.old_indices, self.old_new_ratio),
+                    ],
+                )
+                indices.append(indice)
```
Comment (vinnamkim) [inline src/otx/algo/samplers/class_incremental_sampler.py]:
```suggestion
        for _ in range(self.repeat):
            num_samples = int(self.data_length / (1 + self.old_new_ratio))
            indice = torch.cat(
                    [
                        self.new_indices[torch.randint(0, len(self.new_indices), (num_samples,), generator=generator)],
                        self.old_indices[
                            torch.randint(0, len(self.old_indices), (num_samples * self.old_new_ratio,), generator=generator)
                        ],
                    ],
                )
```

File: src/otx/algo/samplers/class_incremental_sampler.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            num_extra = int(
+                np.ceil(self.data_length * self.repeat / self.batch_size),
+            ) * self.batch_size - len(indices)
+            indices = torch.cat(
+                [
+                    indices,
+                    indices[torch.randint(0, len(indices), (num_extra,), generator=generator)],
+                ],
+            )
+        indices = indices.tolist()
```
Comment (vinnamkim) [inline src/otx/algo/samplers/class_incremental_sampler.py]: Is it necessary? Is there any reason not to keep it as `torch.Tensor`?

Comment (harimkang) [reply src/otx/algo/samplers/class_incremental_sampler.py]: Thank you. Fixed this to use torch.Tensor. - https://github.com/openvinotoolkit/training_extensions/pull/3018/commits/6a1b4e730243addae624727759215803e0a62744

File: src/otx/algo/samplers/class_incremental_sampler.py (hunk 1 above)
Comment (harimkang) [reply src/otx/algo/samplers/class_incremental_sampler.py]: Thanks for the comment. Modify this to create indices separated by batches. - https://github.com/openvinotoolkit/training_extensions/pull/3018/commits/6a1b4e730243addae624727759215803e0a62744

File: src/otx/algo/samplers/class_incremental_sampler.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        dataset: OTXDataset,
+        batch_size: int,
+        old_classes: list[str],
+        new_classes: list[str],
+        efficient_mode: bool = False,
+        num_replicas: int = 1,
+        rank: int = 0,
+        drop_last: bool = False,
+        n_repeats: int = 1,
+        generator: torch.Generator | None = None,
```
Comment (jaegukhyun) [inline src/otx/algo/samplers/class_incremental_sampler.py:55]: Should user specify all of these values in config?

Comment (harimkang) [reply src/otx/algo/samplers/class_incremental_sampler.py:55]: Yes, that's right.
`data.config.train_subset.sampler.init_args` allows the user to set these values.

**Review by jaegukhyun — APPROVED** [review]

**Review by vinnamkim — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Pull request merged**